src/audio/mixin_mixout/mixin_mixout.c

```diff
@@ -90,56 +84,45 @@ struct mixout_data {
 	/* number of currently mixed frames in mixout sink buffer */
 	uint32_t mixed_frames;
 	/*
-	 * Source data is consumed by mixins in mixin_copy() but sink data cannot be
-	 * immediately produced. Sink data is produced by mixout in mixout_copy() after
+	 * Source data is consumed by mixins in mixin_process() but sink data cannot be
+	 * immediately produced. Sink data is produced by mixout in mixout_process() after
 	 * ensuring all connected mixins have mixed their data into mixout sink buffer.
 	 * So for each connected mixin, mixout keeps knowledge of data already consumed
 	 * by mixin but not yet produced in mixout.
 	 */
 	uint32_t pending_frames[MIXOUT_MAX_SOURCES];
 };
 
-static struct comp_dev *mixin_new(const struct comp_driver *drv,
-				  const struct comp_ipc_config *config,
-				  const void *spec)
+static int mixin_init(struct processing_module *mod)
 {
-	struct comp_dev *dev;
+	struct module_data *mod_data = &mod->priv;
+	struct comp_dev *dev = mod->dev;
 	struct mixin_data *md;
 	int i;
 	enum sof_ipc_frame __sparse_cache frame_fmt, valid_fmt;
 
-	comp_cl_dbg(&comp_mixin, "mixin_new()");
-
-	dev = comp_alloc(drv, sizeof(*dev));
-	if (!dev)
-		return NULL;
-
-	dev->ipc_config = *config;
+	comp_dbg(dev, "mixin_init()");
 
 	md = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*md));
-	if (!md) {
-		rfree(dev);
-		return NULL;
-	}
+	if (!md)
+		return -ENOMEM;
 
-	memcpy_s(&md->base_cfg, sizeof(md->base_cfg), spec, sizeof(md->base_cfg));
+	mod_data->private = md;
 
 	for (i = 0; i < MIXIN_MAX_SINKS; i++) {
 		md->sink_config[i].mixer_mode = IPC4_MIXER_NORMAL_MODE;
 		md->sink_config[i].gain = IPC4_MIXIN_UNITY_GAIN;
 	}
 
-	comp_set_drvdata(dev, md);
-
-	audio_stream_fmt_conversion(md->base_cfg.audio_fmt.depth,
-				    md->base_cfg.audio_fmt.valid_bit_depth,
+	audio_stream_fmt_conversion(mod->priv.cfg.base_cfg.audio_fmt.depth,
+				    mod->priv.cfg.base_cfg.audio_fmt.valid_bit_depth,
 				    &frame_fmt, &valid_fmt,
-				    md->base_cfg.audio_fmt.s_type);
+				    mod->priv.cfg.base_cfg.audio_fmt.s_type);
 
 	dev->ipc_config.frame_fmt = frame_fmt;
+	mod->simple_copy = true;
```

Comment:
mixin / mixout can have both multiple sources and sinks, cannot they? Then it cannot be "simple?"

Reply:
No kixin is single source multi sink and mixout is single sink multi source

Reply:
I must either remember something wrong then or somewhere in the code there's a location that implicitly assumes that. Will try to make a note of it next time I come across it (if it really exists)